### PR TITLE
add fill_na, get_imputed_data, replace_negatives and impute_daily_incidence functions

### PR DESCRIPTION
**Pull request by Serena-Wang**

### CODE REVIEW

File: R/get_negative_cases.R [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  locations <- c()
+  dates <- c()
+
+  # Find  observations with negative inc
+  for (i in 1:nrow(data)) {
+    if (data[i, ]$inc < 0) {
+      locations <- c(locations, as.character(data[i, ]$location))
+      dates <- c(dates, as.character(data[i, ]$date))
+    }
+  }
```
Comment (elray1) [inline R/get_negative_cases.R:19]: This could probably be done more easily (and more readable) via calls to `dplyr::filter` and `dplyr::transmute`.  For example, something like

```
return(
  data %>%
    dplyr::filter(inc < 0) %>%
    dplyr::transmute(
      fips = location,
      dates = date
    )
)
```

But also note that for readability it might be nice to keep consistent names (i.e., `location` and `date` rather than `fips` and `dates`), unless there are specific reasons it's advantageous to change the names.

File: R/impute_daily_incidence.R [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#' Imputed incidence value for the single day with reporting anomaly
+#'
+#' @param data time series data for one location in adjustment case
+#' @param adjustment_case a case to adjust
+#' @param measure case or death
+#' @return incidence columns after adjustments
+#' @export
+
```
Comment (elray1) [inline R/impute_daily_incidence.R:8]: Let's keep ROxygen documentation right next to the function definition.

File: R/adjust_daily_incidence.R [hunk 3]
```diff
@@ -0,0 +1,78 @@
+#' Redistribute difference between the original observation and the
+#' imputed incidence value for the single day with reporting anomaly
+#'
+#' @param data time series data for location in adjustment case
+#' @param adjustment_case a case to adjust
+#' @param measure case or death
+#' @return incidence columns after redistribution
+#' @export
+
+
```
Comment (elray1) [inline R/adjust_daily_incidence.R:10]: Let's keep ROxygen documentation right next to the function definition.

File: R/adjust_daily_incidence.R [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#' Redistribute difference between the original observation and the
+#' imputed incidence value for the single day with reporting anomaly
+#'
+#' @param data time series data for location in adjustment case
+#' @param adjustment_case a case to adjust
```
Comment (elray1) [inline R/adjust_daily_incidence.R:5]: Can we add a little more detail describing this parameter?  It looks like the function is expecting this to be a list of length one with one named element, "date"?  If that's true, could we just make this the `adjustment_date`?

File: R/adjust_daily_incidence.R [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#' @param measure case or death
+#' @return incidence columns after redistribution
+#' @export
+
+
+adjust_daily_incidence <- function(data, adjustment_case, seed, measure) {
+  set.seed(seed)
```
Comment (elray1) [inline R/adjust_daily_incidence.R:12]: Is this line (set.seed) still necessary?

File: R/adjust_daily_incidence.R [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  obs <- data[which(data$date == as.Date(adjustment_case$date)), ]$inc
+
+  # Read imputed data
+  replacement <- round(covidData::impute_daily_incidence(data, adjustment_case, measure),
+    digits = 0
+  )
+  # If rep > obs  repl= obs for stan
```
Comment (elray1) [inline R/adjust_daily_incidence.R:20]: This comment seems to be out of date

File: R/fill_na.R [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+#'
+#' @param results a result data.frame
+#' @param adjustments a adjustment case data.frame
+#' @return a data.frame with MAs on adjustment cases
+#' @export
+fill_na <- function(results, adjustments) {
+  adjustments$dates <- as.character(adjustments$dates)
+
+  results %>%
+    dplyr::rowwise() %>%
```
Comment (elray1) [inline R/fill_na.R:11]: Is `rowwise()` necessary here?

File: R/fill_na.R [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    dplyr::mutate(
+      # for each row, if date is in adjustment date and location is in adjustment states
+      inc = ifelse(date %in% adjustments$dates[adjustments$fips == location] |
```
Comment (elray1) [inline R/fill_na.R:14]: I would have expected the condition to check here to be `&` rather than `|`.  Am I missing something?

If this is a bug, we should be sure we have a unit test that can detect it.

File: R/fill_na.R [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        # if first two digits of county fips is the state fips of adjustment states
+        date %in% adjustments$dates[adjustments$fips == stringr::str_sub(location, start = 1, end = 2)], NA_integer_, inc)
```
Comment (elray1) [inline R/fill_na.R:16]: Can we insert some line breaks to get < 80 characters per line?

File: R/get_negative_cases.R [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#' Find all observations with negative incidence in a
+#' given data frame
+#'
+#' @param data data frame
```
Comment (elray1) [inline R/get_negative_cases.R:4]: Let's specify a little more detail about the data frame (what is in it?  what are the columns?)

File: R/jhu_data.R [hunk 11]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
-    
+
     results <- dplyr::bind_rows(results, state_results)
   }
 
   # summarized results for national level
-  if('national' %in% spatial_resolution) {
-    # because we don't filter on states_to_keep as above, we are off by a total
+  if ("national" %in% spatial_resolution) {
+    # because we don"t filter on states_to_keep as above, we are off by a total
```
Comment (elray1) [inline R/jhu_data.R:137]: double quote back to single :)

File: R/jhu_data.R [hunk 12]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+        inc = diff(c(0, cum)),
+        location = "US"
       ) %>%
       dplyr::select(location, date, cum, inc)
 
     results <- dplyr::bind_rows(results, national_results)
   }
+
   
+   if (replace_negatives == TRUE) {
```
Comment (elray1) [inline R/jhu_data.R:153]: This is fine (no need to change), but note you could also just use

```suggestion
   if (replace_negatives) {
```

File: R/jhu_data.R [hunk 13]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
-  # TODO: if temporal_resolution == 'weekly', aggregate daily incidence to weekly incidence here
+
+  if (adjustment_cases != "none" & length(adjustment_cases) > 0) {
+    adjustment_states <- sub("-.*", "", adjustment_cases)
+    adjustment_dates <- sub("^.*?-", "", adjustment_cases)
+    adjustment_state_fips <- unlist(lapply(
+      adjustment_states, function(x) {
+        covidData::fips_codes[which(covidData::fips_codes$abbreviation == x), ]$location
+      }
+    ))
```
Comment (elray1) [inline R/jhu_data.R:168]: `purrr::map_chr` is probably a little cleaner than `unlist(lapply(`:

```suggestion
    adjustment_state_fips <- purrr::map_chr(
      adjustment_states, function(x) {
        covidData::fips_codes[which(covidData::fips_codes$abbreviation == x), ]$location
      }
    )
```

File: R/jhu_data.R [hunk 14]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+      results <- covidData::fill_na(results = results, adjustments = adjustments)
+    }
+
+    
+    if (adjustment_method == "impute_and_redistribute") {
+       if (replace_negatives == FALSE) {
+         results = covidData::replace_negatives(data = results, measure = measure)
+       }
+
+      for (i in 1:nrow(adjustments)) {
```
Comment (elray1) [inline R/jhu_data.R:182]: This is fine in this instance because we've already checked above that there is at least one adjustment, but in general it's good to be in the habit of using `seq_len(nrow(adjustments))`, which is safer: if `nrow(adjustments)` is 0, `1:nrow(adjustments)` would give you the sequence of integers `(1, 0)` so the for loop would run for 2 iterations.  `seq_len(nrow(adjustments))` would run for 0 iterations though.

```suggestion
      for (i in seq_len(nrow(adjustments))) {
```

File: R/jhu_data.R [hunk 15]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+        adjustment_location <- adjustments[i, ]$fips
+        adjustment_date <- as.Date(adjustments[i, ]$date)
+
+        # Get state, counties and national observations for an adjustment case
+        location_data <- results %>%
+          dplyr::filter(stringr::str_sub(location, start = 1, end = 2) %in% adjustment_location |
```
Comment (elray1) [inline R/jhu_data.R:188]: Using comparison with `==` rather than `%in%` might be a little clearer.

```suggestion
          dplyr::filter(stringr::str_sub(location, start = 1, end = 2) == adjustment_location |
```

File: R/jhu_data.R [hunk 16]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+            location == "US" | location == adjustment_location)
+
+        # For each location in data, get imputed data
+        for (fips in unique(location_data$location)) {
+          d <- location_data[location_data$location == fips, ]
+
+          seed <- 1234
```
Comment (elray1) [inline R/jhu_data.R:195]: Do we need the seed here still?

File: R/jhu_data.R [hunk 17]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
   # input: data frame with daily values, output: data frame with approximately 1/7 the # of rows
   # with weekly inc values
   # group_by(location, week), and summarize(inc = sum(inc, na.rm = TRUE))
-  
+  # if weekly temporal resolution, filter to saturdays
+  # TODO: adjust data first, go to weekly resolution later
+  if (temporal_resolution == "weekly") {
+    results <- results %>%
+      dplyr::mutate(sat_date = lubridate::ceiling_date
+      (lubridate::ymd(date), unit = "week") - 1) %>%
```
Comment (elray1) [inline R/jhu_data.R:218]: For readability, it would be helpful to put the paren starting the function call immediately after the function name, then have the line break.

```suggestion
      dplyr::mutate(sat_date = lubridate::ceiling_date(
        lubridate::ymd(date), unit = "week") - 1) %>%
```

File: R/jhu_data.R [hunk 18]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+      dplyr::group_by(location) %>%
+      # if the last week is not complete, drop all obs in that week
+      dplyr::filter(if (max(date) < max(sat_date)) date <= max(sat_date) - 7 else TRUE) %>%
```
Comment (elray1) [inline R/jhu_data.R:221]: Why don't we need to use `ifelse` here?  I believe it could work, just want to think it through.

File: R/jhu_data.R [hunk 19]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+      dplyr::ungroup() %>%
+      # delete date column
+      dplyr::select(-2) %>%
```
Comment (elray1) [inline R/jhu_data.R:224]: Best to specify column names rather than indices in calls to `select()` -- this is safer in case the column ordering ever changes.

File: R/jhu_data.R [hunk 20]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
+      dplyr::rename(date = sat_date) %>%
+      dplyr::group_by(location, date) %>%
+      dplyr::summarize(inc = sum(inc, na.rm = FALSE)) %>%
+      dplyr::ungroup()
+  }
+
   # TODO: aggregate inc to get cum
-  # at this point you'll have a data frame with results for all locations
+  # at this point you"ll have a data frame with results for all locations
```
Comment (elray1) [inline R/jhu_data.R:232]: un-replace double quotes :)

File: R/jhu_data.R [hunk 21]
```diff
@@ -68,102 +73,176 @@ load_jhu_data <- function(
...
   # and either an incorrect cum column that we need to replace (if temporal_resolution == daily)
   # or no cum column  (since previous step got rid of it if temporal_resolution == weekly)
   # group_by(location) [but not week] and summarize(cum = cumsum(inc))
+  # add cumulative col
```
Comment (elray1) [inline R/jhu_data.R:236]: Let's edit this comment to remove my instructions and replace with a description of what it's doing.

File: R/impute_daily_incidence.R [hunk 22]
```diff
@@ -0,0 +1,23 @@
+#' Imputed incidence value for the single day with reporting anomaly
+#'
+#' @param data time series data for one location in adjustment case
+#' @param adjustment_case a case to adjust
+#' @param measure case or death
```
Comment (elray1) [inline R/impute_daily_incidence.R:5]: Let's clarify this documentation a little.  Note that I changed the pluralization here.  If we pluralize, will we break any code?

```suggestion
#' @param measure String specifying measure: "cases" or "deaths"
```

File: R/impute_daily_incidence.R [hunk 23]
```diff
@@ -0,0 +1,23 @@
+#' Imputed incidence value for the single day with reporting anomaly
+#'
+#' @param data time series data for one location in adjustment case
```
Comment (elray1) [inline R/impute_daily_incidence.R:3]: Can we specify what columns the data frame is required to contain?

File: R/impute_daily_incidence.R [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+#' @param adjustment_case a case to adjust
```
Comment (elray1) [inline R/impute_daily_incidence.R:4]: Can we specify format of this argument?

File: R/impute_daily_incidence.R [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+  if (measure == "deaths") {
+    imputed_data <- covidData::jhu_deaths_imputed_data
+  } else {
+    imputed_data <- covidData::jhu_cases_imputed_data
+  }
+
+  location <- unique(data$location)
+  date <- as.Date(adjustment_case$date)
+  
+  unique(imputed_data[which(
```
Comment (elray1) [inline R/impute_daily_incidence.R:19]: Why is there a call to `unique` here?

File: R/get_negative_cases.R [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+
+  # Find  observations with negative inc
+  for (i in 1:nrow(data)) {
+    if (data[i, ]$inc < 0) {
+      locations <- c(locations, as.character(data[i, ]$location))
+      dates <- c(dates, as.character(data[i, ]$date))
+    }
+  }
+
+  data.frame(fips = locations, dates = as.Date(dates))
```
Comment (elray1) [inline R/get_negative_cases.R:21]: Unless there is a specific reason to do otherwise, let's standardize our names around `location` and `date`.

File: R/data.R [hunk 27]
```diff
@@ -15,3 +15,15 @@
 #'
 #' @format A tibble with a row per issue date
 "jhu_cases_data"
+
+
+#' imputed deaths
+#'
+#' @format A data.frame
```
Comment (elray1) [inline R/data.R:22]: Let's add a little more detail to this description: what are the variables in the data frame?

File: R/data.R [hunk 28]
```diff
@@ -15,3 +15,15 @@
...
+"jhu_deaths_imputed_data"
+
+#' imputed cases
+#'
+#' @format A data.frame
```
Comment (elray1) [inline R/data.R:27]: Let's add a little more detail to this description: what are the variables in the data frame?

File: R/replace_negatives.R [hunk 29]
```diff
@@ -0,0 +1,35 @@
+#' Replace negative incidence with imputed data and redistribute
+#' residuals across observations before adjustment date
+#'
+#' @param data data frame
+#' @param measure death or case
+#' @return data frame after replacement
```
Comment (elray1) [inline R/replace_negatives.R:6]: Let's add a little more detail describing these things

File: R/replace_negatives.R [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+#'
+#' @export
+#'
+replace_negatives <- function(data, measure) {
+
+  # find observations with negative inc
+  adjustments <- covidData::get_negative_cases(data)
+
+  # replace negative incidents with imputed data for all replated locations
```
Comment (elray1) [inline R/replace_negatives.R:15]: Minor spelling note: "incidence" :)

```suggestion
  # replace negative incidence with imputed data for all replated locations
```

File: R/replace_negatives.R [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+  for (i in 1:nrow(adjustments)) {
```
Comment (elray1) [inline R/replace_negatives.R:16]: We probably need `seq_len(nrow(adjustments))` here in case there are 0 adjustments to make (i.e., no observations of negative incidence)?  If this is a bug, we should make sure we have a unit test that detects it.

```suggestion
  for (i in seq_len(nrow(adjustments))) {
```

File: R/replace_negatives.R [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+    case <- adjustments[i, ]
+    seed <- 1234
```
Comment (elray1) [inline R/replace_negatives.R:18]: Do we need a seed here?

File: R/replace_negatives.R [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+    adjustment_location <- case$fips
+    
+    # get county, state and national locations
+    location_data <- data %>%
+      dplyr::filter(location == stringr::str_sub(adjustment_location, start = 1, end = 2) |
+                      location == "US" | location == adjustment_location) 
+    
+    for (fips in unique(location_data$location)) {
+
```
Comment (elray1) [inline R/replace_negatives.R:27]: deletable line

```suggestion
```

File: R/replace_negatives.R [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+      imputed_data <- covidData::adjust_daily_incidence(data[data$location == fips, ], case, seed, measure)
```
Comment (elray1) [inline R/replace_negatives.R:28]: Let's make sure this is < 80 characters

```suggestion
      imputed_data <- covidData::adjust_daily_incidence(
        data[data$location == fips, ], case, seed, measure)
```

File: code/data-processing/get_imputed_data.R [hunk 35]
```diff
@@ -0,0 +1,187 @@
...
+    if (data[i, ]$inc < 0) {
+      locations <- c(locations, as.character(data[i, ]$location))
+      dates <- c(dates, as.character(data[i, ]$date))
+    }
+  }
+  
+  adjustments <- data.frame(fips = locations, dates = as.Date(dates))
+  
+  return(adjustments)
+}
```
Comment (elray1) [inline code/data-processing/get_imputed_data.R:22]: I know we went back and forth about this, but it seems like we ended up adding this function to the package as an exported function.  If that's a decision we're good with, we can delete it from here.

```suggestion
```

File: tests/testthat/run-tests.R [hunk 36]
```diff
@@ -0,0 +1,2 @@
+library(devtools)
+devtools::test()
```
Comment (elray1) [inline tests/testthat/run-tests.R:2]: add new line to end of file

```suggestion
devtools::test()

```

File: tests/testthat/test-replace-negatives.R [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+testthat::context("replace_negatives")
+
+# read in data
+data <- covidData::load_jhu_data(spatial_resolution = "state", 
+                     temporal_resolution = "daily",
+                     measure = "deaths", replace_negatives = FALSE, 
+                     adjustment_cases = "none") %>%
+  # CO
+  dplyr:: filter(location == "08")
+
```
Comment (elray1) [inline tests/testthat/test-replace-negatives.R:11]: Let's add a second example test case where there are not negative reports to replace.

File: tests/testthat/test-adjust-daily-incidence.R [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+
+
+# case b, 2
+# observed is positive, replacement also positive but less than observed
+test_that("positive diff, incidents are unchanged after the adjustment date", {
+  # 2020-04-24
+
+  expect_true(all(new_inc_case_b[data$date > "2020-04-24"] ==
+    data[data$date > "2020-04-24", ]$inc))
+})
```
Comment (elray1) [inline tests/testthat/test-adjust-daily-incidence.R:68]: Let's also test that cumulative counts are unchanged on and after the adjustment date, for both case a and case b.

File: R/jhu_data.R [hunk 39]
```diff
@@ -4,61 +4,66 @@
...
+#' of "CO-2020-04-24", "MS-2020-06-22", "DE-2020-06-23", "NJ-2020-06-25". These
 #' refer to locations and times affected by reporting anomalies documented at
 #' https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data#user-content-retrospective-reporting-of-probable-cases-and-deaths
 #' @param adjustment_method string specifying how anomalies are adjusted.
-#' 'fill_na' will replace affected observations with NAs.
-#' Currently the only option is 'impute_inc', which leaves cumulative counts
+#' "fill_na" will replace affected observations with NAs and correct daily
+#' cumulative counts for all dates on and after adjustment date.
+#' Currently the only option is "impute_inc", which leaves cumulative counts
 #' unaffected and imputes the incidence value with the 7-day mean incidence
```
Comment (elray1) [inline R/jhu_data.R:22]: Is this description of handling of cumulative counts still correct?

Comment (elray1) [reply R/jhu_data.R:22]: Also is this now called "impute_and_redistribute"?

File: R/impute_daily_incidence.R (hunk 25 above)
Comment (elray1) [reply R/impute_daily_incidence.R:19]: This is not a bug, but let's remove `unique` from here and instead ensure that at the time we generate imputed values we save only a single imputed value for each combination of location and date. It seems like we could do this by inserting a check of whether the combination of `fips` and `adjustment_date` is already in the `results` data frame at https://github.com/reichlab/covidData/blob/fill-nas/code/data-processing/get_imputed_data.R#L125

File: tests/testthat/test-adjust-daily-incidence.R (hunk 38 above)
Comment (elray1) [reply tests/testthat/test-adjust-daily-incidence.R:68]: Maybe these tests should go in a different file, `test-load_jhu_data-impute_and_redistribute.R`

File: R/jhu_data.R (hunk 15 above)
Comment (Serena-Wang) [reply R/jhu_data.R]: This is for getting state-level data when adjustment_location is a county, and also the other way around. So it's only matching the first two digits (state part) of a fips code.
